openedx/core/djangoapps/credentials_service/models.py

```diff
@@ -1,3 +1,264 @@
+    )
+
+    class Meta(object):
+        unique_together = (('course_id', 'certificate_type', 'site'),)
+
+    def __unicode__(self):
+        """Unicode representation. """
+        return '{course_id}, {certificate_type}'.format(
+            course_id=self.course_id,
+            certificate_type=self.certificate_type,
+        )
+
+
+class ProgramCertificate(AbstractCertificate):
+    """
+    Configuration for Program Certificates.
+    """
+    program_id = models.PositiveIntegerField(db_index=True)
+
+    def __unicode__(self):
+        """Unicode representation. """
+        return '{program_id}'.format(
+            program_id=self.program_id
+        )
+
+
+class UserCredential(TimeStampedModel):
+    """
+    Credentials issued to a learner.
+    """
+    AWARDED, REVOKED = (
+        'awarded', 'revoked',
+    )
+
+    STATUSES_CHOICES = (
+        (AWARDED, _('awarded')),
+        (REVOKED, _('revoked')),
+    )
+
+    credential_content_type = models.ForeignKey(ContentType)
+    credential_id = models.PositiveIntegerField()
+    credential = GenericForeignKey('credential_content_type', 'credential_id')
+
+    username = models.CharField(max_length=255, db_index=True)
+    status = models.CharField(
+        max_length=255,
+        choices=STATUSES_CHOICES,
+        default=AWARDED
+    )
+    download_url = models.CharField(
+        max_length=255, blank=True, null=True,
+        help_text=_('Download URL for the PDFs.')
+    )
+    uuid = models.UUIDField(default=uuid.uuid4, editable=False)
+
+    class Meta(object):
+        unique_together = (('username', 'credential_content_type', 'credential_id'),)
+
+    def __unicode__(self):
+        """Unicode representation. """
+        return '{username}, {status}'.format(
+            username=self.username,
+            status=self.status
+        )
+
+
+class UserCredentialAttribute(TimeStampedModel):
+    """
+    Different attributes of User's Credential such as white list, grade etc.
+    """
+    user_credential = models.ForeignKey(UserCredential)
+    namespace = models.CharField(max_length=255)
+    name = models.CharField(max_length=255)
+    value = models.CharField(max_length=255)
+
+    def __unicode__(self):
+        """Unicode representation. """
+        return '{user_credential}, {namespace}, {name}'.format(
+            user_credential=self.user_credential,
+            namespace=self.namespace,
+            name=self.name
+        )
+
+
+class CertificateTemplateAsset(TimeStampedModel):
+    """
+    Certificate Template Asset model to add content files for a certificate
+    template.
+    """
+    name = models.CharField(max_length=255)
+    asset_file = models.FileField(upload_to=template_assets_path)
+
+    def __unicode__(self):
+        """Unicode representation. """
+        return '{name}'.format(
+            name=self.name
+        )
+
+    def save(self, *args, **kwargs):
+        """Save the certificate template asset."""
+        if self.pk is None:
```

Comment:
I don't think the code you have written will work as expected. After the first save, you set `asset_file` but you never call `super(...).save(...)` a second time. Remove the `else`. Also, add a docstring explaining what you're doing.